### PR TITLE
HPCC-20910 Prevent slave initialization aborting too early

### PR DESCRIPTION
**Pull request by jakesmith**
Extend timeout for activity initialization time and make it
configurable, plus log why waiting in logs.
Prior to this change, if a query was blocked e.g. due to a file
locking issue in the master during initialization, it would cause
the slaves to fail after timeout (25 mins).
Not it will by default continue for up to 2hrs, configurable
via global/workunit option "actInitWaitTimeMins"

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
NB: this is sort of a bug fix.. there was a change between 5.6.8 and 6.0 that meant a long waiting blocked query, would timeout here, with what was a confusing error message.
In 5.6.8 it would block on the lock before it ever got here and would have never timed out.

There's a separate issue of indefinite unconfigurable blocking on file locks, I've opened a separate JIRA for that (HPCC-20912). i.e. it would be sensible if any Thor query that is in file lookup that is blocked for a hours causes an error by default.

@mckellyln - please review.

File: thorlcr/graph/thgraphslave.hpp [hunk 1]
```diff
@@ -474,6 +474,7 @@ class graphslave_decl CJobSlave : public CJobBase
     Owned<IPropertyTree> workUnitInfo;
     size32_t oldNodeCacheMem;
     unsigned channelMemoryMB;
+    unsigned actInitWaitTimeMins = 0;
```
Comment (mckellyln) [inline thorlcr/graph/thgraphslave.hpp]: Doesn't really matter, but why not initialize this to DEFAULT_MAX_ACTINITWAITTIME ?

File: thorlcr/graph/thgraphslave.cpp [hunk 2]
```diff
@@ -949,8 +949,18 @@ bool CSlaveGraph::recvActivityInitData(size32_t parentExtractSz, const byte *par
 
     if (syncInitData())
     {
-        if (!graphCancelHandler.recv(queryJobChannel().queryJobComm(), msg, 0, mpTag, NULL, LONGTIMEOUT))
-            throw MakeStringException(0, "Error receiving actinit data for graph: %" GIDPF "d", graphId);
+        CTimeMon timer;
+        while (!graphCancelHandler.recv(queryJobChannel().queryJobComm(), msg, 0, mpTag, NULL, MEDIUMTIMEOUT))
+        {
+            if (graphCancelHandler.isCancelled())
+                throw MakeStringException(0, "Aborted whilst receiving actinit data for graph: %" GIDPF "d", graphId);
```
Comment (mckellyln) [inline thorlcr/graph/thgraphslave.cpp]: Aborted whilst waiting to receive actinit data ?

File: thorlcr/graph/thgraphslave.cpp [hunk 3]
```diff
@@ -949,8 +949,18 @@ bool CSlaveGraph::recvActivityInitData(size32_t parentExtractSz, const byte *par
 
     if (syncInitData())
     {
-        if (!graphCancelHandler.recv(queryJobChannel().queryJobComm(), msg, 0, mpTag, NULL, LONGTIMEOUT))
-            throw MakeStringException(0, "Error receiving actinit data for graph: %" GIDPF "d", graphId);
+        CTimeMon timer;
+        while (!graphCancelHandler.recv(queryJobChannel().queryJobComm(), msg, 0, mpTag, NULL, MEDIUMTIMEOUT))
```
Comment (mckellyln) [inline thorlcr/graph/thgraphslave.cpp]: If checking for isCancelled in loop, why do it at MEDIUMTIMEOUT (30000) interval ?  Why not at 1000 or so to end without delay ?

**Comment by mckellyln** [comment]
@jakesmith looks good, just some minor comments/questions.

File: thorlcr/graph/thgraphslave.cpp (hunk 3 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.cpp]: The loop is here mainly to issue a line of tracing every interval, and we wouldn't want that every 1 second. If there's an abort condition it won't wait that long, the recv will be cancelled.

File: thorlcr/graph/thgraphslave.cpp (hunk 2 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.cpp]: I'll change.

File: thorlcr/graph/thgraphslave.hpp (hunk 1 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.hpp]: could do, I'll change.

**Comment by jakesmith** [comment]
@richardkchapman - shall I retarget this to candidate 7.0.4? 

@mckellyln - made 2 minor changes mentioned.

File: thorlcr/graph/thgraphslave.hpp (hunk 1 above)
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.hpp]: @mckellyln - NB: also changed this to DEFAULT_MAX_ACTINITWAITTIME_MINS and corrected constant as mentioned here: https://github.com/hpcc-systems/HPCC-Platform/pull/11903#discussion_r235661190

File: thorlcr/graph/thgraphslave.cpp (hunk 3 above)
Comment (mckellyln) [reply thorlcr/graph/thgraphslave.cpp]: ok.  I like small granularity for knowing it was cancelled, and then tracing only every 30 iterations, etc.  But no matter, just my approach.

**Comment by mckellyln** [comment]
@jakesmith looks good

**Review by mckellyln — APPROVED** [review]

**Pull request merged**